### ISSUE TITLE
Add room key import & export from file

### ISSUE THREAD

**Issue by jahway603** [krille-chan/fluffychat#942]
### Feature Description

Add room key import & export from file. This is currently missing.

### Rationale

This is a required feature for a user to successfully move to or from FluffyChat to or from another Matrix protocol client.

**Comment by AppTester7** [comment]
This is a feature that is really required here for anyone to have their own personal e2ee backup on their own devices
- It should be noted that the element app features/functions:

- 'Connect this session to Key Backup'
- 'Set up Secure Backup'

Do NOT work properly at all
Both functions 'assume' that you already have Security Key in the first place - for reasons that are beyond me
Plus even in the past, when i have have had a security key - there were still issues around it

Testing:

PC: Linux Desktop client app Element
Android: Fluffychat

My emoji verification was successfull at both ends, however this did not let my PC client decrypt any messages
I then teyed out the fluffy web chat on my PC, i successfully completed the emoji verification with my android fluffy chat app, but i could still not decrypt any messages in fluffy web on my PC


FURTHER INFO

**Comment by AppTester7** [comment]
[Element - Security Key Purpose, Use Case, Creation, Backup & Restoration Tests - No Personal Data.txt](https://github.com/krille-chan/fluffychat/files/15147745/Element.-.Security.Key.Purpose.Use.Case.Creation.Backup.Restoration.Tests.-.No.Personal.Data.txt)

**Comment by jahway603** [comment]
@krille-chan have you see this issue yet?

**Comment by jahway603** [comment]
Adding the same comment as was added in Issue https://github.com/famedly/matrix-dart-sdk/issues/1785:

In the "early days" of using the Matrix protocol, in 2017-2018, exporting and importing room keys was the ONLY method available to see all messages on every Matrix client session, so you had to do this between all of them.

Fast forward to today, where this online key backup system is never working correctly, and I STILL USE the above 2017-2018 method because it ALWAYS WORKS. FluffyChat is the ONLY client which I am prohibited from using this method as it has not yet been implemented.

Please understand that just because you're not using this feature does not mean that it is not important to others. Again, I will state that this is one of the most important issues for Matrix client interoperability.

**Comment by jahway603** [comment]
@krille-chan I just thought of yet another use case for adding the Matrix protocol key import & export from file feature into matrix-dart-sdk (Issue https://github.com/famedly/matrix-dart-sdk/issues/1785) and ultimately into FluffyChat:

A user can export their room keys to an Offline File before a server they are using is **Permanently Shutdown**. Then after they create a new account on a totally different server, then can import that Offline File and when they re-join the rooms they were in, they will see all of the encrypted messages that they had Offline keys saved for. This use case will NEVER be able to use the "online key backup with SSSS" functionality...

I know this is a valid use case as I have personally done this in the past before privacytools.io shut down their Matrix server for good.
What do you think @AppTester7  ?

**Comment by AppTester7** [comment]
Im still 100% in favour of manual e2ee export/import keys
Why should there be an absolute and complete 'dependency' on *any* server to store the keys ?
Besides which i have made complaints over the official keys server functionality not working 99% of the time
Issues regarding this have been wide-spread, and i have not seen much improvement in this e2ee server keys feature
(you know, for things like verification between 2 clients via the emoji test)
Manual e2ee key export/import can help to by pass these limitations when there are issues with server/client communications.

As for Element-X, well that entire concept is not a reality yet, i recently went into the element-x room, and i was told that hardly any severs actually had yet implemented the option for the fast message service

Im not going to lose all my messages again using fluffy chat, untill the devs implement e2ee export/import functionality

**Comment by jahway603** [comment]
This issue still exists in the latest Version 1.26.1 of Fluffychat on Android.

**Comment by jahway603** [comment]
This issue still exists in the latest Version 2.1.0 of Fluffychat on Android.

**Comment by twassman** [comment]
Seems to be a duplicate of #124?

**Comment by jahway603** [comment]
> Seems to be a duplicate of [#124](https://github.com/krille-chan/fluffychat/issues/124)?

It does appear this may be a duplicate, but Github closed that other issue.

**Comment by CEbbinghaus** [comment]
I can't believe this issue is still open 2 years later. Fluffychat is somehow the only client that works while simultaneously missing this core feature. Hopefully we can get it added sometime soon.

**Comment by krille-chan** [comment]
> I can't believe this issue is still open 2 years later. Fluffychat is somehow the only client that works while simultaneously missing this core feature. Hopefully we can get it added sometime soon.

It's in your hands! You can donate here: https://ko-fi.com/krille and add a message that you would like to have this feature. If enough money comes together, this would increase the chance 😊

**Comment by CEbbinghaus** [comment]
I presume contributions are also welcome towards this? Strapped for both cash and time but time less than money. Been working around it by using Shildi for the key export and import and relying on the dehydrated device to sync the keys within accounts (in particular I'm struggling with encryption keys missing when joining group chats).

**Comment by krille-chan** [comment]
> I presume contributions are also welcome towards this? Strapped for both cash and time but time less than money. Been working around it by using Shildi for the key export and import and relying on the dehydrated device to sync the keys within accounts (in particular I'm struggling with encryption keys missing when joining group chats).

Yes but would first need an implementation plan. I'd suggest the following:

Implement a new extension to the Matrix Dart SDK with two methods: `importMegolmSessions()` and `exportMegolmSessions()` (while exportMegolmSessions() should download everything from the online key backup if available beforehand). Those methods must also have unit tests.

Second step then in FluffyChat would be to create a new page in settings where the user comes on when clicking on the list tile with "Chat Backup" in the main settings page. There should be buttons to export/import the key backup in a three-dot-menu at the bottom right while in the content of the page there should be a SwitchListTile telling if it is enabled and below (if enabled) a red button to reset the recovery key

**Comment by CEbbinghaus** [comment]
> (while exportMegolmSessions() should download everything from the online key backup if available beforehand)

Does FluffyChat not already do this? And if not is there a way to force it to? Based on behavior I'd hazard the guess that it does not automatically download all keys given the amount of times I have to click on an encrypted message before it retrieves the relevwnt key.

Would probably cut out ~60% of the usecase if the keys sync with the dehydrated device worked properly

**Comment by krille-chan** [comment]
> > (while exportMegolmSessions() should download everything from the online key backup if available beforehand)
> 
> Does FluffyChat not already do this? And if not is there a way to force it to? Based on behavior I'd hazard the guess that it does not automatically download all keys given the amount of times I have to click on an encrypted message before it retrieves the relevwnt key.
> 
> Would probably cut out ~60% of the usecase if the keys sync with the dehydrated device worked properly

No FluffyChat only downloads the key when an encrypted message appeares on the screen. It doesn't download the whole key backup in advance.

**Comment by CEbbinghaus** [comment]
> No FluffyChat only downloads the key when an encrypted message appeares on the screen. It doesn't download the whole key backup in advance.

My observations are actually that it does not download the key unless I click onto the encrypted message itself. In any case I think a better first step would be adding a button to fluffychat to download all keys, That way it is easy to tell at which point the key export/import is actually required due to some keys having not properly synced. Does that sound reasonable?

Should I create a separate issue for that?

**Comment by jahway603** [comment]
@CEbbinghaus there is further discussion (which ended up being locked) about this issue in the matrix-dart-sdk at https://github.com/famedly/matrix-dart-sdk/issues/1785

**Comment by berlincount** [comment]
I'm affected by this issue today. (Re-)Pushing existing keys to the SSSS (after recovery phrase change) doesn't seem to be happening, either, so there is no recourse for me - old content on channels previous joined/read with FluffyChat stays unreadable 😓

**Comment by krille-chan** [comment]
Please stay on topic. I'll lock the conversation for now
